File: src/secure-write.ts
import fs from 'fs';

export interface WriteSecureOptions {
  mode: number; // e.g. 0o600
}

/**
 * Atomically write a file and set its mode. Write to `{target}.tmp`,
 * fsync, rename over the target, then chmod. Used for any file that
 * holds a secret (OAuth tokens, state files with session metadata).
 *
 * On any failure the temp file is removed and the exception propagates.
 */
export function writeFileSecure(
  target: string,
  content: string | Buffer,
  opts: WriteSecureOptions,
): void {
  if (target.includes('\x00')) {
    throw new Error('writeFileSecure: target contains NUL byte');
  }
  const tmp = `${target}.tmp`;
  let fd: number | null = null;
  try {
    fd = fs.openSync(tmp, 'w', opts.mode);
    fs.writeSync(fd, content as any);
    fs.fsyncSync(fd);
    fs.closeSync(fd);
    fd = null;
    fs.renameSync(tmp, target);
    // Belt-and-braces: chmod after rename in case an fs layer ignored mode on open
    fs.chmodSync(target, opts.mode);
  } catch (err) {
    if (fd !== null) {
      try {
        fs.closeSync(fd);
      } catch {
        /* already closed */
      }
    }
    try {
      fs.unlinkSync(tmp);
    } catch {
      /* tmp may not exist */
    }
    throw err;
  }
}

File: scripts/sync/email_ingest/secure_write.py
"""Atomic secure file writes for tokens and state files.

Write to `{target}.tmp` with mode 0o600, fsync, rename over target.
Used for any file holding a secret (OAuth tokens, state files).
"""

import os
from pathlib import Path
from typing import Union


def write_file_secure(
    target: Union[str, Path],
    content: Union[str, bytes],
    *,
    mode: int = 0o600,
) -> None:
    target = Path(target)
    tmp = target.with_name(target.name + ".tmp")

    fd = None
    try:
        # O_WRONLY | O_CREAT | O_TRUNC, create with requested mode
        fd = os.open(
            tmp,
            os.O_WRONLY | os.O_CREAT | os.O_TRUNC,
            mode,
        )
        if isinstance(content, str):
            data = content.encode("utf-8")
        else:
            data = content
        os.write(fd, data)
        os.fsync(fd)
        os.close(fd)
        fd = None

        os.replace(tmp, target)
        # Belt-and-braces chmod in case O_CREAT mode was filtered by umask
        os.chmod(target, mode)
    except Exception:
        if fd is not None:
            try:
                os.close(fd)
            except OSError:
                pass
        try:
            tmp.unlink()
        except FileNotFoundError:
            pass
        raise
